### ISSUE TITLE
[BUG]: Run compare_models but always return empty list

### ISSUE THREAD

**Issue by Haminh630** [pycaret/pycaret#3690]
### pycaret version checks

- [X] I have checked that this issue has not already been reported [here](https://github.com/pycaret/pycaret/issues).

- [X] I have confirmed this bug exists on the [latest version](https://github.com/pycaret/pycaret/releases) of pycaret.

- [X] I have confirmed this bug exists on the master branch of pycaret (pip install -U git+https://github.com/pycaret/pycaret.git@master).

### Issue Description

Hello everyone,

I always face the issue that after run compare_models() return the empty list

How can I help me to fix this bug?
<img width="1242" alt="Screenshot 2023-08-06 at 08 30 27" src="https://github.com/pycaret/pycaret/assets/141295366/2064ff20-e382-4e68-a0c2-41f73ecd7000">

<img width="1242" alt="Screenshot 2023-08-06 at 08 30 44" src="https://github.com/pycaret/pycaret/assets/141295366/e37e3d5a-e231-4aa7-ad68-6b94ad17e426">

<img width="1242" alt="Screenshot 2023-08-06 at 08 31 09" src="https://github.com/pycaret/pycaret/assets/141295366/e4c9a2aa-dfae-4799-a3ad-382fd54552d5">

### Reproducible Example

```python
import numpy as np
import pandas as pd

train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")

valid_df = train[train['month'] == '2022-11'].reset_index(drop = True)
train_df = train[train['month'] != '2022-11'].reset_index(drop = True)

features = [col for col in train_df.columns if col not in ["userID_hash", "month", "y"]]

y = 'y'

# import the regression module**
from pycaret.regression import *

# initialize setup**
s = setup(data = train_df, test_data = valid_df, target = y, fold_strategy = 'timeseries', 
          numeric_features = features, fold = 3, transform_target = True, session_id = 123, index=False)


best = compare_models(sort = 'MAE', include = ['lr', 'en'])

finalzed_model = finalize_model(best)
preds = predict_model(finalzed_model, data=test)
```

### Expected Behavior

I hope I know how can I get best model from compare_models() function

### Actual Results

```python-traceback
Output exceeds the size limit. Open the full output data in a text editor
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[30], line 1
----> 1 prediction_holdout = predict_model(best);

File ~/anaconda3/envs/time_series_dl/lib/python3.8/site-packages/pycaret/regression/functional.py:1903, in predict_model(estimator, data, round, verbose)
   1900 if experiment is None:
   1901     experiment = _EXPERIMENT_CLASS()
-> 1903 return experiment.predict_model(
   1904     estimator=estimator,
   1905     data=data,
   1906     round=round,
   1907     verbose=verbose,
   1908 )

File ~/anaconda3/envs/time_series_dl/lib/python3.8/site-packages/pycaret/regression/oop.py:2194, in RegressionExperiment.predict_model(self, estimator, data, round, verbose)
   2140 def predict_model(
   2141     self,
   2142     estimator,
   (...)
   2145     verbose: bool = True,
   2146 ) -> pd.DataFrame:
   2147     """
   2148     This function predicts ``Label`` using a trained model. When ``data`` is
   2149     None, it predicts label on the holdout set.
...
   4956 pred = pipeline.inverse_transform(pred)
   4957 # Need to convert labels back to numbers
   4958 # TODO optimize

AttributeError: 'list' object has no attribute 'predict'
```

### Installed Versions

<details>
3.0.4
</details>

**Comment by Umar-cs** [comment]
Same err
StreamlitAPIException: ("Could not convert 'PassengerId' with type str: tried to convert to int64", 'Conversion failed for column Value with type object')
solved this err using "C:\Users\your_user\anaconda3\envs\your_env_name\Lib\site-packages\streamlit\config.py"
In the config.py locate dataFrameSerialization = "arrow"
instead of "arrow" change to "legacy" --> save
but the problem is that now getting compare_df output empty.
need some help to resolve this issue.

`if choice == "Modelling": 
    chosen_target = st.selectbox('Choose the Target Column', df.columns)
    if st.button('Run Modelling'): 
        #silent=True
        setup(df, target=chosen_target, verbose = False)
        setup_df = pull()
        st.dataframe(setup_df)
        best_model = compare_models()
        compare_df = pull()
        st.info("model")
        st.dataframe(compare_df)
        best_model
        save_model(best_model, 'best_model')
        `
![Screenshot 2023-08-06 160922](https://github.com/pycaret/pycaret/assets/82081986/be5aecc3-a08f-4083-a132-61a435f0c37c)

**Comment by moezali1** [comment]
@Haminh630 `compare_models` return empty list when model training fails. If you set `errors = 'raise'` in `compare_models` you will be able to see what is causing the failure. 

Also if you can share the `train.csv` and `test.csv` here so I can troubleshoot what is causing this. Thanks

**Comment by mukeshvadapali72** [comment]
I am also getting the empty list and when used error = 'raise' I am getting this error - TerminatedWorkerError: A worker process managed by the executor was unexpectedly terminated. This could be caused by a segmentation fault while calling the function or by an excessive memory usage causing the Operating System to kill the worker.

**Comment by moezali1** [comment]
@mukeshvadapali72 How big is your dataset? Can you also try this by passing `n_jobs = 1` and let us know if that works?

**Comment by Umar-cs** [comment]
Used **errors = 'raise'** in **compare_models**  got this ValueError: n_splits=10 cannot be greater than the number of members in each class. Kindly guide me to resolve this error

**Comment by ngupta23** [comment]
Please google this error message. This is coming from scikit since you have too few classes for the number of folds. You can try reducing the number of folds to run.

![image](https://github.com/pycaret/pycaret/assets/33585645/7e6aa0b5-d0e4-40f3-88cf-3e552bb26151)

**Comment by huangliang0828** [comment]
I also  meet the  similar problem , not only  Binary Classification Tutorial Level Beginner (中文) - CLF101.ipynb, but also for other .ipynb files from Simplifying Machine Learning with PyCaret-book(Classification.ipynb, and even myself files) 
all the .ipynb files are trusted.
compared_models(fold=3) # The results of step automatically disappears around 3s (as shown in the figure) [lighGBM] [info]... [lighGBM] [info] ..., as if this step did not run, there is no result output of compared_models()
conda install python=3.8.16
pip install pycaret==2.3.10 
pip check  # Env No broken requirements found.
![image](https://github.com/pycaret/pycaret/assets/52817734/6fb87555-d66d-4c31-87d6-0d968b95741b)
compared_models(fold=3) # the results automatically disappears.  
![image](https://github.com/pycaret/pycaret/assets/52817734/2cd5e69f-2fba-42c3-b4da-38e80aaa825b)

**Comment by huangliang0828** [comment]
for the Binary Classification Tutorial Level Beginner (中文) - CLF101.ipynb
![image](https://github.com/pycaret/pycaret/assets/52817734/5ac04569-38d0-4152-9b09-251cca7ecba0)
--------------------------------------------------------------------------
AttributeError         Traceback (most recent call last)
Cell In[8], line 1
----> 1 best_model = compare_models(fold = 3,n_select=10,errors = 'raise') # errors = 'raise'
      2 # compare_models?
File D:\ProgramData\miniconda3\envs\pycaret\lib\site-packages\pycaret\classification.py:823, in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, experiment_custom_tags, probability_threshold, verbose, display, parallel)
    820     del params["parallel"]
    821     return parallel.compare_models(compare_models, params)
--> 823 return pycaret.internal.tabular.compare_models(
    824     include=include,
    825     exclude=exclude,
    826     fold=fold,
    827     round=round,
    828     cross_validation=cross_validation,
    829     sort=sort,
    830     n_select=n_select,
    831     budget_time=budget_time,
    832     turbo=turbo,
    833     errors=errors,
    834     fit_kwargs=fit_kwargs,
    835     groups=groups,
    836     experiment_custom_tags=experiment_custom_tags,
    837     probability_threshold=probability_threshold,
    838     verbose=verbose,
    839     display=display,
    840 )
File D:\ProgramData\miniconda3\envs\pycaret\lib\site-packages\pycaret\internal\tabular.py:2199, in compare_models(include, exclude, fold, round, cross_validation, sort, n_select, budget_time, turbo, errors, fit_kwargs, groups, experiment_custom_tags, probability_threshold, verbose, display)
   2184 create_model_args = dict(
   2185     estimator=model,
   2186     system=False,
   (...)
   2196     refit=False,
   2197 )
   2198 if errors == "raise":
-> 2199     model, model_fit_time = create_model_supervised(**create_model_args)
   2200     model_results = pull(pop=True)
   2201 else:

File D:\ProgramData\miniconda3\envs\pycaret\lib\site-packages\pycaret\internal\tabular.py:3316, in create_model_supervised(estimator, fold, round, cross_validation, predict, fit_kwargs, groups, refit, verbose, system, X_train_data, y_train_data, metrics, experiment_custom_tags, add_to_model_list, probability_threshold, display, return_train_score, **kwargs)
   3312     indices = ["Mean"]
   3313 model_results = color_df(
   3314     model_results, "yellow", indices, axis=1
   3315 )
-> 3316 model_results = model_results.set_precision(round)
   3318 # end runtime
   3319 runtime_end = time.time()

AttributeError: 'Styler' object has no attribute 'set_precision'

**Comment by moezali1** [comment]
@huangliang0828 Can you please try this with PyCaret >3.0 and let us know if you are still facing the issue.

**Comment by ashleydavyUHBW** [comment]
> @mukeshvadapali72 How big is your dataset? Can you also try this by passing `n_jobs = 1` and let us know if that works?

I got the same error as above user. I'm trying to get time-series tutorial to run, with the test/tutorial data ('airline') only having 144 rows. With my original data (not tutorial data, but again only 597 long, with 1 column) it would take about 5 minutes to run, but then would be pointless as nothing would return.
I added in n_jobs = 1 in the setup and added include = ['auto_arima','exp_smooth'] which then did run, but I didn't get the output of a grid/table with various columns showing data accuracy etc.

If I used the OOP api then the result returned:
![image](https://github.com/pycaret/pycaret/assets/142324069/cb5c1b9b-5496-4aa8-9274-056ecdde8999)


If I do the functional api, then the n_jobs=1 has at least made it work and best is not empty, it contains type ExponentialSmoothing. However the grid doesn't display comparing the models
![image](https://github.com/pycaret/pycaret/assets/142324069/0896e3f0-3e5b-40c6-84b3-20d1393a815e)

However when following the **tutorial for classification** using diabetes dataset, the below output grid did generate as expected, 
![image](https://github.com/pycaret/pycaret/assets/142324069/84edcacd-522e-41b1-8c4b-855367672810)

When the Time-series is running I do get a visible output for the first chunk of dots:
![image](https://github.com/pycaret/pycaret/assets/142324069/61af0baa-6058-47ff-9793-4f99d8dfc45e)
But then when finished, it all clears and never shows the grid, unlike when I ran it with the classification.

I'm new to all of this, and installed pyCaret into a new venv this morning, so definitely on latest version of PyCaret! (Just checked and it's 3.0.4)

**Comment by ashleydavyUHBW** [comment]
tl;dr - I'm using Visual Studio Code and clicking "clear all outputs" before running currently seems to largely fix the issue.

After looking at this last night, I had left code running, and it had frozen on 20% for a long time, decided to terminate after maybe 30 minutes
Again this morning,  ran some code including just a few estimators and it worked! I tried adding in and taking out some and thought that adding in polytrend made it work, but this turned out to be a red-herring (though I _think_ I've successfully found that 'auto_arima' causes it to hang (possibly) indefinitely. Maybe it is just a bit slower to run (like all the '_cds_dt' models).

Ultimately I discovered adding in more or less or different models (using include = []) initially appeared to make a difference, but it didn't. Clicking "Clear all outputs" worked!

Interestingly, now when running (with exclude = ['auto_arima]), the first 19 models appear in the table, with the progress bar being on 97% before it then disappeared and has continued working on the the remaining models ['rf_cds_dt', 'et_cds_dt', 'gbr_cds_dt', 'ada_cds_dt','lightgbm_cds_dt'], but without the progress bar. However the top box still appeared saying the status and estimator, so I could see when it was nearing the end due to having pasted the list of models().index in another cell.

**Comment by tpduarte** [comment]
> @huangliang0828 Can you please try this with PyCaret >3.0 and let us know if you are still facing the issue.

I had a very similar error and installing pycaret 3.1.0 version solved it.

**Comment by ashleydavyUHBW** [comment]
> > @huangliang0828 Can you please try this with PyCaret >3.0 and let us know if you are still facing the issue.
> 
> I had a very similar error and installing pycaret 3.1.0 version solved it.

As I'm relatively new to python generally - do you have a suggestion of how best to do this?
I guess also ensuring that the version if of PyCaret I update is in the correct place?

I have Python 3.7.9, and tried installing pyCaret but it failed to import, which lead me to learn through StackOverflow that it's best to create a virtual environment and then have PyCaret be the first thing that is installed, so it has all the correct dependencies.

I did that, but as I had not created a venv before and didn't know that giving it a name with spaces in would create multiple venv's, I've ended up with the venv with PyCaret installed into being called "by" 
![image](https://github.com/pycaret/pycaret/assets/142324069/736f02a2-f086-404b-b2a0-02210dec399f)

I've now looked and seen that this site - https://pypi.org/project/pycaret/  - says python >3.8 is tested, so thought I should start again in a venv not based on 3.7 
although I note it worked in 3.7 and that same site shows 3.7 at the top:
![image](https://github.com/pycaret/pycaret/assets/142324069/299e91a1-c889-4aac-a4fc-5b95264e3bf5)

To summarize/restate my question simply:
**What's the quickest/best way to update/uninstall PyCaret?**
Currently installing it into my new 3.10 venv has taken 13 minutes and counting😩

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
